packages/@lwc/engine-core/src/framework/hot-swaps.ts

```diff
@@ -199,14 +199,28 @@ export function swapComponent(
     newComponent: LightningElementConstructor
 ): boolean {
     if (process.env.NODE_ENV !== 'production') {
-        if (isComponentConstructor(oldComponent) && isComponentConstructor(newComponent)) {
+        const isOldCtorAComponent = isComponentConstructor(oldComponent);
+        const isNewCtorAComponent = isComponentConstructor(newComponent);
+        if (isOldCtorAComponent && isNewCtorAComponent) {
             swappedComponentMap.set(oldComponent, newComponent);
             return rehydrateHotComponent(oldComponent);
+        } else if (isOldCtorAComponent === false && isNewCtorAComponent === true) {
+            throw new TypeError(
+                `Invalid Component: Attempting to swap a non-component with a component`
+            );
+        } else if (isOldCtorAComponent === true && isNewCtorAComponent === false) {
+            throw new TypeError(
+                `Invalid Component: Attempting to swap a component with a non-component`
+            );
         } else {
-            throw new TypeError(`Invalid Component`);
+            // The dev-server relies on the presence of registerComponent() as a way to determine a
+            // component module. However, the compiler cannot definitively add registerComponent()
+            // transformation only to a component constructor. Hence the dev-server may attempt to
+            // hot swap javascript modules that look like a component and should not cause the app
+            // to fail. To allow that, this api ignores such hot swap attempts.
+            return false;
         }
     }
```

Comment:
We should be able to test this logic in Karma:

https://github.com/salesforce/lwc/blob/6bed69892783c28f21ed4f25d205d5437eeb209f/packages/%40lwc/integration-karma/test/swapping/components/index.spec.js#L12

Reply:
✅